File: CitySimulator/Models/Building.swift
//
//  Building.swift
//  CitySimulator
//
//  Created by AI进化论-花生 on 2024/01/01.
//

import Foundation

/// 建筑类型枚举
enum BuildingType: String, CaseIterable, Codable {
    // 住宅建筑
    case house = "house"
    case apartment = "apartment"
    case villa = "villa"
    case skyscraper = "skyscraper"
    
    // 商业建筑
    case shop = "shop"
    case supermarket = "supermarket"
    case mall = "mall"
    case restaurant = "restaurant"
    case hotel = "hotel"
    
    // 工业建筑
    case lumberMill = "lumberMill"
    case quarry = "quarry"
    case steelMill = "steelMill"
    case foodFactory = "foodFactory"
    case powerPlant = "powerPlant"
    
    // 公共设施
    case school = "school"
    case hospital = "hospital"
    case policeStation = "policeStation"
    case fireStation = "fireStation"
    case park = "park"
    case library = "library"
    
    // 交通设施
    case road = "road"
    case bridge = "bridge"
    case busStop = "busStop"
    case subwayStation = "subwayStation"
    
    /// 获取建筑类型的中文名称
    var displayName: String {
        switch self {
        case .house: return "普通住宅"
        case .apartment: return "公寓楼"
        case .villa: return "别墅"
        case .skyscraper: return "摩天大楼"
        case .shop: return "商店"
        case .supermarket: return "超市"
        case .mall: return "购物中心"
        case .restaurant: return "餐厅"
        case .hotel: return "酒店"
        case .lumberMill: return "木材厂"
        case .quarry: return "采石场"
        case .steelMill: return "钢铁厂"
        case .foodFactory: return "食品加工厂"
        case .powerPlant: return "发电厂"
        case .school: return "学校"
        case .hospital: return "医院"
        case .policeStation: return "警察局"
        case .fireStation: return "消防局"
        case .park: return "公园"
        case .library: return "图书馆"
        case .road: return "道路"
        case .bridge: return "桥梁"
        case .busStop: return "公交站"
        case .subwayStation: return "地铁站"
        }
    }
    
    /// 获取建筑类型分类
    var category: BuildingCategory {
        switch self {
        case .house, .apartment, .villa, .skyscraper:
            return .residential
        case .shop, .supermarket, .mall, .restaurant, .hotel:
            return .commercial
        case .lumberMill, .quarry, .steelMill, .foodFactory, .powerPlant:
            return .industrial
        case .school, .hospital, .policeStation, .fireStation, .park, .library:
            return .public
        case .road, .bridge, .busStop, .subwayStation:
            return .transportation
        }
    }
    
    /// 获取建筑解锁等级
    var unlockLevel: Int {
        switch self {
        case .house: return 1
        case .apartment: return 3
        case .villa: return 8
        case .skyscraper: return 15
        case .shop: return 2
        case .supermarket: return 5
        case .mall: return 12
        case .restaurant: return 4
        case .hotel: return 10
        case .lumberMill: return 3
        case .quarry: return 3
        case .steelMill: return 6
        case .foodFactory: return 5
        case .powerPlant: return 8
        case .school: return 4
        case .hospital: return 6
        case .policeStation: return 5
        case .fireStation: return 5
        case .park: return 2
        case .library: return 7
        case .road: return 1
        case .bridge: return 9
        case .busStop: return 6
        case .subwayStation: return 12
        }
    }
}

/// 建筑分类枚举
enum BuildingCategory: String, CaseIterable {
    case residential = "residential"
    case commercial = "commercial"
    case industrial = "industrial"
    case public = "public"
    case transportation = "transportation"
    
    var displayName: String {
        switch self {
        case .residential: return "住宅区"
        case .commercial: return "商业区"
        case .industrial: return "工业区"
        case .public: return "公共设施"
        case .transportation: return "交通设施"
        }
    }
}

/// 建筑数据模型
struct Building: Codable, Identifiable {
    let id = UUID()
    var type: BuildingType
    var level: Int
    var position: GridPosition
    var buildTime: Date
    var lastUpgradeTime: Date?
    var lastMaintenanceTime: Date
    
    // 建筑属性
    var capacity: Int = 0
    var income: Double = 0.0
    var maintenanceCost: Double = 0.0
    var happinessBonus: Double = 0.0
    var pollutionLevel: Double = 0.0
    
    // 资源生产/消耗
    var productionRate: [ResourceType: Double]?
    var consumptionRate: [ResourceType: Double]?
    
    // 建筑状态
    var isUnderConstruction: Bool = false
    var isMaintenanceRequired: Bool = false
    var isUpgrading: Bool = false
    
    init(type: BuildingType, level: Int = 1, position: GridPosition) {
        self.type = type
        self.level = level
        self.position = position
        self.buildTime = Date()
        self.lastMaintenanceTime = Date()
        
        // 根据建筑类型设置初始属性
        updateBuildingProperties()
    }
    
    /// 更新建筑属性
    mutating func updateBuildingProperties() {
        switch type {
        case .house:
            capacity = 4 * level
            income = 0.0
            maintenanceCost = 10.0 * Double(level)
            happinessBonus = 5.0 * Double(level)
            
        case .apartment:
            capacity = 8 * level
            income = 0.0
            maintenanceCost = 15.0 * Double(level)
            happinessBonus = 3.0 * Double(level)
            
        case .villa:
            capacity = 2 * level
            income = 0.0
            maintenanceCost = 25.0 * Double(level)
            happinessBonus = 10.0 * Double(level)
            
        case .skyscraper:
            capacity = 20 * level
            income = 0.0
            maintenanceCost = 50.0 * Double(level)
            happinessBonus = 2.0 * Double(level)
            
        case .shop:
            capacity = 0
            income = 50.0 * Double(level)
            maintenanceCost = 20.0 * Double(level)
            happinessBonus = 2.0 * Double(level)
            
        case .supermarket:
            capacity = 0
            income = 100.0 * Double(level)
            maintenanceCost = 30.0 * Double(level)
            happinessBonus = 3.0 * Double(level)
            
        case .mall:
            capacity = 0
            income = 300.0 * Double(level)
            maintenanceCost = 80.0 * Double(level)
            happinessBonus = 5.0 * Double(level)
            
        case .restaurant:
            capacity = 0
            income = 80.0 * Double(level)
            maintenanceCost = 25.0 * Double(level)
            happinessBonus = 4.0 * Double(level)
            
        case .hotel:
            capacity = 0
            income = 200.0 * Double(level)
            maintenanceCost = 60.0 * Double(level)
            happinessBonus = 3.0 * Double(level)
            
        case .lumberMill:
            capacity = 0
            income = 0.0
            maintenanceCost = 40.0 * Double(level)
            happinessBonus = -1.0 * Double(level)
            pollutionLevel = 2.0 * Double(level)
            productionRate = [.wood: 10.0 * Double(level)]
            
        case .quarry:
            capacity = 0
            income = 0.0
            maintenanceCost = 35.0 * Double(level)
            happinessBonus = -1.0 * Double(level)
            pollutionLevel = 2.0 * Double(level)
            productionRate = [.stone: 8.0 * Double(level)]
            
        case .steelMill:
            capacity = 0
            income = 0.0
            maintenanceCost = 60.0 * Double(level)
            happinessBonus = -2.0 * Double(level)
            pollutionLevel = 3.0 * Double(level)
            productionRate = [.steel: 5.0 * Double(level)]
            
        case .foodFactory:
            capacity = 0
            income = 0.0
            maintenanceCost = 30.0 * Double(level)
            happinessBonus = 1.0 * Double(level)
            pollutionLevel = 1.0 * Double(level)
            productionRate = [.food: 15.0 * Double(level)]
            
        case .powerPlant:
            capacity = 0
            income = 0.0
            maintenanceCost = 100.0 * Double(level)
            happinessBonus = -3.0 * Double(level)
            pollutionLevel = 5.0 * Double(level)
            productionRate = [.energy: 20.0 * Double(level)]
            
        case .school:
            capacity = 0
            income = 0.0
            maintenanceCost = 50.0 * Double(level)
            happinessBonus = 3.0 * Double(level)
            
        case .hospital:
            capacity = 0
            income = 0.0
            maintenanceCost = 80.0 * Double(level)
            happinessBonus = 4.0 * Double(level)
            
        case .policeStation:
            capacity = 0
            income = 0.0
            maintenanceCost = 60.0 * Double(level)
            happinessBonus = 2.0 * Double(level)
            
        case .fireStation:
            capacity = 0
            income = 0.0
            maintenanceCost = 70.0 * Double(level)
            happinessBonus = 2.0 * Double(level)
            
        case .park:
            capacity = 0
            income = 0.0
            maintenanceCost = 20.0 * Double(level)
            happinessBonus = 6.0 * Double(level)
            
        case .library:
            capacity = 0
            income = 0.0
            maintenanceCost = 40.0 * Double(level)
            happinessBonus = 3.0 * Double(level)
            
        case .road:
            capacity = 0
            income = 0.0
            maintenanceCost = 5.0 * Double(level)
            happinessBonus = 1.0 * Double(level)
            
        case .bridge:
            capacity = 0
            income = 0.0
            maintenanceCost = 15.0 * Double(level)
            happinessBonus = 1.0 * Double(level)
            
        case .busStop:
            capacity = 0
            income = 0.0
            maintenanceCost = 10.0 * Double(level)
            happinessBonus = 2.0 * Double(level)
            
        case .subwayStation:
            capacity = 0
            income = 0.0
            maintenanceCost = 30.0 * Double(level)
            happinessBonus = 3.0 * Double(level)
        }
    }
    
    /// 获取建筑建造成本
    func getBuildCost() -> [ResourceType: Double] {
        switch type {
        case .house:
            return [.wood: 50, .stone: 30, .steel: 10]
        case .apartment:
            return [.wood: 80, .stone: 60, .steel: 20]
        case .villa:
            return [.wood: 120, .stone: 100, .steel: 40]
        case .skyscraper:
            return [.wood: 200, .stone: 300, .steel: 150]
        case .shop:
            return [.wood: 40, .stone: 20, .steel: 5]
        case .supermarket:
            return [.wood: 80, .stone: 60, .steel: 20]
        case .mall:
            return [.wood: 150, .stone: 200, .steel: 100]
        case .restaurant:
            return [.wood: 60, .stone: 40, .steel: 15]
        case .hotel:
            return [.wood: 100, .stone: 150, .steel: 80]
        case .lumberMill:
            return [.wood: 30, .stone: 50, .steel: 20]
        case .quarry:
            return [.wood: 20, .stone: 10, .steel: 30]
        case .steelMill:
            return [.wood: 40, .stone: 80, .steel: 50]
        case .foodFactory:
            return [.wood: 50, .stone: 30, .steel: 25]
        case .powerPlant:
            return [.wood: 100, .stone: 200, .steel: 300]
        case .school:
            return [.wood: 80, .stone: 100, .steel: 40]
        case .hospital:
            return [.wood: 120, .stone: 150, .steel: 80]
        case .policeStation:
            return [.wood: 60, .stone: 80, .steel: 50]
        case .fireStation:
            return [.wood: 70, .stone: 90, .steel: 60]
        case .park:
            return [.wood: 20, .stone: 10, .steel: 5]
        case .library:
            return [.wood: 60, .stone: 80, .steel: 30]
        case .road:
            return [.wood: 5, .stone: 10, .steel: 2]
        case .bridge:
            return [.wood: 30, .stone: 50, .steel: 40]
        case .busStop:
            return [.wood: 15, .stone: 20, .steel: 10]
        case .subwayStation:
            return [.wood: 50, .stone: 100, .steel: 80]
        }
    }
    
    /// 获取建筑升级成本
    func getUpgradeCost() -> [ResourceType: Double] {
        let baseCost = getBuildCost()
        let multiplier = Double(level + 1)
        return baseCost.mapValues { $0 * multiplier }
    }
    
    /// 获取建筑升级时间（秒）
    func getUpgradeTime() -> TimeInterval {
        let baseTime: TimeInterval = 300 // 5分钟
        return baseTime * Double(level)
    }
    
    /// 检查是否需要维护
    func needsMaintenance() -> Bool {
        let maintenanceInterval: TimeInterval = 3600 * 24 * 7 // 7天
        return Date().timeIntervalSince(lastMaintenanceTime) > maintenanceInterval
    }
}

/// 网格位置结构
struct GridPosition: Codable, Equatable {
    var x: Int
    var y: Int
    
    init(x: Int, y: Int) {
        self.x = x
        self.y = y
    }
}

File: CitySimulatorAndroid/app/src/main/java/com/citysimulator/game/data/model/Building.kt
package com.citysimulator.game.data.model

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import com.citysimulator.game.data.converter.DateConverter
import java.util.Date

/**
 * 建筑类型枚举
 * 
 * 定义游戏中所有可建造的建筑类型，按功能分类。
 */
enum class BuildingType {
    // 住宅类建筑
    HOUSE,           // 普通住宅
    APARTMENT,       // 公寓
    VILLA,           // 别墅
    SKYSCRAPER,      // 摩天大楼
    
    // 商业类建筑
    SHOP,            // 商店
    SUPERMARKET,     // 超市
    MALL,            // 商场
    RESTAURANT,      // 餐厅
    HOTEL,           // 酒店
    
    // 工业生产建筑
    FARM,            // 农场 - 生产粮食
    LUMBER_MILL,     // 木材厂 - 生产木材
    QUARRY,          // 采石场 - 生产石料
    STEEL_MILL,      // 钢铁厂 - 生产钢材
    FOOD_FACTORY,    // 食品厂 - 加工食品
    BAKERY,          // 面包房 - 用粮食生产高级食物
    SAWMILL,         // 锯木厂 - 用木材生产板材
    POWER_PLANT,     // 发电厂
    
    // 公共设施
    SCHOOL,          // 学校
    HOSPITAL,        // 医院
    POLICE_STATION,  // 警察局
    FIRE_STATION,    // 消防局
    PARK,            // 公园 - 提升繁荣度
    LIBRARY,         // 图书馆 - 提升教育水平
    
    // 交通设施
    ROAD,            // 道路
    BRIDGE,          // 桥梁
    BUS_STOP,        // 公交站
    SUBWAY_STATION,  // 地铁站
    
    // 工业建筑
    FACTORY,         // 工厂
    
    // 能源建筑
    SOLAR_PLANT,     // 太阳能发电厂
    WIND_FARM,       // 风力发电厂
    
    // 环保建筑
    RECYCLING_CENTER,      // 回收中心
    WASTE_MANAGEMENT,      // 垃圾管理中心
    
    // 水利设施
    WATER_TOWER,           // 水塔
    WATER_TREATMENT_PLANT, // 水处理厂
    
    // 金融建筑
    BANK,                  // 银行
    
    // 办公建筑
    OFFICE,                // 办公楼
    
    // 体育建筑
    STADIUM,               // 体育场
    
    // 智能建筑
    SMART_CENTER,    // 智慧中心
    
    // 教育建筑
    UNIVERSITY,      // 大学
    
    // 文化建筑
    MUSEUM,          // 博物馆
    THEATER,         // 剧院
    
    // 高级建筑
    SPACE_CENTER,    // 太空中心
    AI_CENTER;       // AI中心
    
    /**
     * 获取建筑类型的显示名称
     */
    fun getDisplayName(): String {
        return when (this) {
            HOUSE -> "住宅"
            APARTMENT -> "公寓"
            VILLA -> "别墅"
            SKYSCRAPER -> "摩天大楼"
            SHOP -> "商店"
            SUPERMARKET -> "超市"
            MALL -> "商场"
            RESTAURANT -> "餐厅"
            HOTEL -> "酒店"
            FARM -> "农场"
            LUMBER_MILL -> "木材厂"
            QUARRY -> "采石场"
            STEEL_MILL -> "钢铁厂"
            FOOD_FACTORY -> "食品厂"
            BAKERY -> "面包房"
            SAWMILL -> "锯木厂"
            POWER_PLANT -> "发电厂"
            SCHOOL -> "学校"
            HOSPITAL -> "医院"
            POLICE_STATION -> "警察局"
            FIRE_STATION -> "消防局"
            PARK -> "公园"
            LIBRARY -> "图书馆"
            ROAD -> "道路"
            BRIDGE -> "桥梁"
            BUS_STOP -> "公交站"
            SUBWAY_STATION -> "地铁站"
            FACTORY -> "工厂"
            SOLAR_PLANT -> "太阳能发电厂"
            WIND_FARM -> "风力发电厂"
            RECYCLING_CENTER -> "回收中心"
            WASTE_MANAGEMENT -> "垃圾管理中心"
            WATER_TOWER -> "水塔"
            WATER_TREATMENT_PLANT -> "水处理厂"
            BANK -> "银行"
            OFFICE -> "办公楼"
            STADIUM -> "体育场"
            SMART_CENTER -> "智慧中心"
            UNIVERSITY -> "大学"
            MUSEUM -> "博物馆"
            THEATER -> "剧院"
            SPACE_CENTER -> "太空中心"
            AI_CENTER -> "AI中心"
        }
    }
}

/**
 * 建筑类别枚举
 * 
 * 将建筑按功能进行分类，便于管理和显示。
 */
enum class BuildingCategory {
    RESIDENTIAL,     // 住宅
    COMMERCIAL,      // 商业
    INDUSTRIAL,      // 工业
    PUBLIC,          // 公共设施
    TRANSPORTATION   // 交通设施
}

/**
 * 建筑状态枚举
 * 
 * 定义建筑可能处于的各种状态。
 */
enum class BuildingStatus {
    NORMAL,          // 正常
    CONSTRUCTION,    // 建设中
    UPGRADING,       // 升级中
    MAINTENANCE,     // 需要维护
    DAMAGED          // 损坏
}

/**
 * 建筑位置数据类
 * 
 * 表示建筑在城市网格中的位置。
 */
data class BuildingPosition(
    val x: Int,
    val y: Int
)

/**
 * 建筑数据模型
 * 
 * 代表城市中的建筑物，包含建筑的基本信息和状态。
 * 使用Room数据库进行持久化存储。
 * 
 * @property id 建筑唯一标识符
 * @property type 建筑类型
 * @property level 建筑等级
 * @property position 建筑位置
 * @property buildTime 建造时间
 * @property status 建筑状态
 * @property isUnderConstruction 是否正在建造
 * @property isUpgrading 是否正在升级
 * @property isMaintenanceRequired 是否需要维护
 * @property lastMaintenanceDate 最后维护日期
 * @property efficiency 建筑效率
 * @property capacity 建筑容量
 * @property income 建筑收入
 * @property maintenanceCost 维护成本
 * 
 * @author AI进化论-花生
 * @since 1.0
 */
@Entity(tableName = "buildings")
@TypeConverters(DateConverter::class)
data class Building(
    @PrimaryKey
    val id: String,
    val type: BuildingType,
    val level: Int = 1,
    @Embedded(prefix = "position_")
    val position: BuildingPosition,
    val buildTime: Date = Date(),
    val status: BuildingStatus = BuildingStatus.NORMAL,
    val isUnderConstruction: Boolean = false,
    val isUpgrading: Boolean = false,
    val isMaintenanceRequired: Boolean = false,
    val lastMaintenanceDate: Date = Date(),
    val efficiency: Float = 1.0f,
    val capacity: Int = 0,
    val income: Int = 0,
    val maintenanceCost: Int = 0,
    val customName: String? = null  // 自定义名称，用于简化建筑系统
) {
    /**
     * 获取建筑类别
     * 
     * @return 建筑所属类别
     */
    fun getCategory(): BuildingCategory {
        return when (type) {
            BuildingType.HOUSE, BuildingType.APARTMENT, BuildingType.VILLA, BuildingType.SKYSCRAPER -> BuildingCategory.RESIDENTIAL
            BuildingType.SHOP, BuildingType.SUPERMARKET, BuildingType.MALL, BuildingType.RESTAURANT, BuildingType.HOTEL, BuildingType.BANK, BuildingType.OFFICE -> BuildingCategory.COMMERCIAL
            BuildingType.FARM, BuildingType.LUMBER_MILL, BuildingType.QUARRY, BuildingType.STEEL_MILL, BuildingType.FOOD_FACTORY, BuildingType.BAKERY, BuildingType.SAWMILL, BuildingType.POWER_PLANT, BuildingType.FACTORY, BuildingType.SOLAR_PLANT, BuildingType.WIND_FARM, BuildingType.RECYCLING_CENTER, BuildingType.WASTE_MANAGEMENT, BuildingType.WATER_TOWER, BuildingType.WATER_TREATMENT_PLANT -> BuildingCategory.INDUSTRIAL
            BuildingType.SCHOOL, BuildingType.HOSPITAL, BuildingType.POLICE_STATION, BuildingType.FIRE_STATION, BuildingType.PARK, BuildingType.LIBRARY, BuildingType.UNIVERSITY, BuildingType.MUSEUM, BuildingType.THEATER, BuildingType.SMART_CENTER, BuildingType.SPACE_CENTER, BuildingType.AI_CENTER, BuildingType.STADIUM -> BuildingCategory.PUBLIC
            BuildingType.ROAD, BuildingType.BRIDGE, BuildingType.BUS_STOP, BuildingType.SUBWAY_STATION -> BuildingCategory.TRANSPORTATION
        }
    }
    
    /**
     * 获取建筑显示名称
     * 
     * @return 建筑的中文显示名称（优先使用自定义名称）
     */
    fun getDisplayName(): String {
        // 如果有自定义名称，优先使用
        if (!customName.isNullOrEmpty()) {
            return customName
        }
        
        // 否则使用BuildingType的默认名称
        val defaultName = when (type) {
            BuildingType.HOUSE -> "住宅"
            BuildingType.APARTMENT -> "公寓"
            BuildingType.VILLA -> "别墅"
            BuildingType.SKYSCRAPER -> "摩天大楼"
            BuildingType.SHOP -> "商店"
            BuildingType.SUPERMARKET -> "超市"
            BuildingType.MALL -> "商场"
            BuildingType.RESTAURANT -> "餐厅"
            BuildingType.HOTEL -> "酒店"
            BuildingType.FARM -> "农场"
            BuildingType.LUMBER_MILL -> "木材厂"
            BuildingType.QUARRY -> "采石场"
            BuildingType.STEEL_MILL -> "钢铁厂"
            BuildingType.FOOD_FACTORY -> "食品厂"
            BuildingType.BAKERY -> "面包房"
            BuildingType.SAWMILL -> "锯木厂"
            BuildingType.POWER_PLANT -> "发电厂"
            BuildingType.SCHOOL -> "学校"
            BuildingType.HOSPITAL -> "医院"
            BuildingType.POLICE_STATION -> "警察局"
            BuildingType.FIRE_STATION -> "消防局"
            BuildingType.PARK -> "公园"
            BuildingType.LIBRARY -> "图书馆"
            BuildingType.ROAD -> "道路"
            BuildingType.BRIDGE -> "桥梁"
            BuildingType.BUS_STOP -> "公交站"
            BuildingType.SUBWAY_STATION -> "地铁站"
            BuildingType.FACTORY -> "工厂"
            BuildingType.SOLAR_PLANT -> "太阳能发电厂"
            BuildingType.WIND_FARM -> "风力发电厂"
            BuildingType.RECYCLING_CENTER -> "回收中心"
            BuildingType.SMART_CENTER -> "智慧中心"
            BuildingType.UNIVERSITY -> "大学"
            BuildingType.MUSEUM -> "博物馆"
            BuildingType.THEATER -> "剧院"
            BuildingType.SPACE_CENTER -> "太空中心"
            BuildingType.AI_CENTER -> "AI中心"
            BuildingType.WASTE_MANAGEMENT -> "垃圾管理中心"
            BuildingType.WATER_TOWER -> "水塔"
            BuildingType.WATER_TREATMENT_PLANT -> "水处理厂"
            BuildingType.BANK -> "银行"
            BuildingType.OFFICE -> "办公楼"
            BuildingType.STADIUM -> "体育场"
        }
        return defaultName
    }
    
    /**
     * 获取建筑描述
     * 
     * @return 建筑的功能描述
     */
    fun getDescription(): String {
        return when (type) {
            BuildingType.HOUSE -> "为居民提供住所，增加城市人口"
            BuildingType.APARTMENT -> "高效的住宅建筑，容纳更多居民"
            BuildingType.VILLA -> "豪华住宅，提升居民幸福感"
            BuildingType.SKYSCRAPER -> "现代化高层建筑，容纳大量居民"
            BuildingType.SHOP -> "提供日常商品，增加城市收入"
            BuildingType.SUPERMARKET -> "大型购物场所，提升商业收入"
            BuildingType.MALL -> "综合性商业中心，大幅增加收入"
            BuildingType.RESTAURANT -> "提供餐饮服务，增加收入"
            BuildingType.HOTEL -> "接待游客，增加旅游收入"
            BuildingType.FARM -> "生产粮食资源，供应城市食物"
            BuildingType.LUMBER_MILL -> "生产木材资源"
            BuildingType.QUARRY -> "开采石材资源"
            BuildingType.STEEL_MILL -> "生产钢材资源"
            BuildingType.FOOD_FACTORY -> "生产食品资源"
            BuildingType.BAKERY -> "将粮食加工成高级食物，提升繁荣度"
            BuildingType.SAWMILL -> "将木材加工成板材，用于建造高级建筑"
            BuildingType.POWER_PLANT -> "为城市提供电力"
            BuildingType.SCHOOL -> "提升居民教育水平"
            BuildingType.HOSPITAL -> "提供医疗服务，提升健康指数"
            BuildingType.POLICE_STATION -> "维护城市安全"
            BuildingType.FIRE_STATION -> "提供消防服务"
            BuildingType.PARK -> "提升环境质量和居民幸福感"
            BuildingType.LIBRARY -> "提升居民教育水平"
            BuildingType.ROAD -> "连接城市各个区域"
            BuildingType.BRIDGE -> "跨越河流和障碍"
            BuildingType.BUS_STOP -> "提供公共交通服务"
            BuildingType.SUBWAY_STATION -> "高效的地铁交通"
            BuildingType.FACTORY -> "生产工业产品，增加收入"
            BuildingType.SOLAR_PLANT -> "清洁能源发电，减少污染"
            BuildingType.WIND_FARM -> "风力发电，环保节能"
            BuildingType.RECYCLING_CENTER -> "处理废物，保护环境"
            BuildingType.SMART_CENTER -> "智慧城市管理，提升效率"
            BuildingType.UNIVERSITY -> "高等教育机构，培养人才"
            BuildingType.MUSEUM -> "文化展示，提升城市品位"
            BuildingType.THEATER -> "艺术表演，丰富文化生活"
            BuildingType.SPACE_CENTER -> "太空科技，引领未来"
            BuildingType.AI_CENTER -> "人工智能，智能管理"
            BuildingType.WASTE_MANAGEMENT -> "垃圾管理中心，处理废弃物"
            BuildingType.WATER_TOWER -> "水塔，提供城市供水"
            BuildingType.WATER_TREATMENT_PLANT -> "水处理厂，净化供水"
            BuildingType.BANK -> "银行，提供金融服务"
            BuildingType.OFFICE -> "办公楼，提供商业办公空间"
            BuildingType.STADIUM -> "体育场，举办体育赛事"
        }
    }
    
    /**
     * 获取建筑图标资源ID
     * 
     * @return 图标资源ID
     */
    fun getIconResourceId(): String {
        return when (type) {
            BuildingType.HOUSE -> "ic_house"
            BuildingType.APARTMENT -> "ic_apartment"
            BuildingType.VILLA -> "ic_villa"
            BuildingType.SKYSCRAPER -> "ic_skyscraper"
            BuildingType.SHOP -> "ic_shop"
            BuildingType.SUPERMARKET -> "ic_supermarket"
            BuildingType.MALL -> "ic_mall"
            BuildingType.RESTAURANT -> "ic_restaurant"
            BuildingType.HOTEL -> "ic_hotel"
            BuildingType.FARM -> "ic_farm"
            BuildingType.LUMBER_MILL -> "ic_lumber_mill"
            BuildingType.QUARRY -> "ic_quarry"
            BuildingType.STEEL_MILL -> "ic_steel_mill"
            BuildingType.FOOD_FACTORY -> "ic_food_factory"
            BuildingType.BAKERY -> "ic_bakery"
            BuildingType.SAWMILL -> "ic_sawmill"
            BuildingType.POWER_PLANT -> "ic_power_plant"
            BuildingType.SCHOOL -> "ic_school"
            BuildingType.HOSPITAL -> "ic_hospital"
            BuildingType.POLICE_STATION -> "ic_police_station"
            BuildingType.FIRE_STATION -> "ic_fire_station"
            BuildingType.PARK -> "ic_park"
            BuildingType.LIBRARY -> "ic_library"
            BuildingType.ROAD -> "ic_road"
            BuildingType.BRIDGE -> "ic_bridge"
            BuildingType.BUS_STOP -> "ic_bus_stop"
            BuildingType.SUBWAY_STATION -> "ic_subway_station"
            BuildingType.FACTORY -> "ic_factory"
            BuildingType.SOLAR_PLANT -> "ic_solar_plant"
            BuildingType.WIND_FARM -> "ic_wind_farm"
            BuildingType.RECYCLING_CENTER -> "ic_recycling_center"
            BuildingType.SMART_CENTER -> "ic_smart_center"
            BuildingType.UNIVERSITY -> "ic_university"
            BuildingType.MUSEUM -> "ic_museum"
            BuildingType.THEATER -> "ic_theater"
            BuildingType.SPACE_CENTER -> "ic_space_center"
            BuildingType.AI_CENTER -> "ic_ai_center"
            BuildingType.WASTE_MANAGEMENT -> "ic_waste_management"
            BuildingType.WATER_TOWER -> "ic_water_tower"
            BuildingType.WATER_TREATMENT_PLANT -> "ic_water_treatment"
            BuildingType.BANK -> "ic_bank"
            BuildingType.OFFICE -> "ic_office"
            BuildingType.STADIUM -> "ic_stadium"
        }
    }
    
    /**
     * 获取升级所需时间（秒）
     * 
     * @return 升级所需时间
     */
    fun getUpgradeTime(): Long {
        return level * 3600L // 每级增加1小时
    }
    
    /**
     * 获取维护间隔时间（秒）
     * 
     * @return 维护间隔时间
     */
    fun getMaintenanceInterval(): Long {
        return 7 * 24 * 3600L // 7天
    }
    
    /**
     * 检查是否需要维护
     * 
     * @return 是否需要维护
     */
    fun needsMaintenance(): Boolean {
        val now = Date()
        val timeSinceLastMaintenance = now.time - lastMaintenanceDate.time
        return timeSinceLastMaintenance > getMaintenanceInterval()
    }
}
